addons/account/models/account_payment.py

```diff
@@ -309,6 +309,20 @@ def open_payment_matching_screen(self):
             'context': action_context,
         }
 
+    @api.onchange('amount')
+    def _onchange_amount(self):
+        res = {}
+        journal_type = ['bank', 'cash']
+        if self.amount == 0:
+            journal_type.append('general')
+            self.payment_difference_handling = 'reconcile'
+            self.journal_id = self.env['account.journal'].search([('type', '=', 'bank')], limit=1)
+        res['domain'] = {'journal_id': [
+            ('type', 'in', journal_type),
+            self.payment_type == 'inbound' and ('at_least_one_inbound', '=', True) or ('at_least_one_outbound', '=', True)
+        ]}
+        return res
```

Comment:
@smetl   setting domain in xml would not be effective, because on onchange of payment_type, domain for journal is set to bank and cash  see here: https://github.com/odoo/odoo/blob/master/addons/account/models/account_payment.py#L346